### PR TITLE
Crud actions

### PR DESCRIPTION
**Pull request by dannyvassallo**
I don't know if the tests are exactly what you're looking for.

### CODE REVIEW

File: app/controllers/lists_controller.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
+class ListsController < ApplicationController
+  before_action :authenticate_user!
+
+  def show
+    @lists = current_user.lists
```
Comment (eliotsykes) [inline app/controllers/lists_controller.rb:10]: Line 5 looks like it'd belong in an `index` action as its multiple `List`s, not a single list.

File: app/controllers/lists_controller.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def new
+    @list = List.new
+  end
+
+  def edit
+    @list = List.find(params[:id])
```
Comment (eliotsykes) [inline app/controllers/lists_controller.rb:20]: With a little URL tampering, a user can edit another user's lists currently.

Write a feature spec that tests:
- a logged in attacker cannot view and edit another users list

Write a controller spec that tests:
- a logged in attacker cannot delete another users list

Once the specs are failing in expected ways, then modify the `edit`, `update`, and `destroy` actions to be more secure.

File: app/controllers/lists_controller.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  end
+
+  def create
+    @list = current_user.lists.build(list_params)
+
+    if @list.save
+      flash[:notice] = "List was saved."
+      redirect_to [@list]
```
Comment (eliotsykes) [inline app/controllers/lists_controller.rb]: The array surrounding `@list` can be safely omitted.

File: app/controllers/lists_controller.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      render :new
+    end
+  end
+
+  def update
+    @list = List.find(params[:id])
+
+    if @list.update_attributes(list_params)
+      flash[:notice] = "List was updated."
+      redirect_to [@list]
```
Comment (eliotsykes) [inline app/controllers/lists_controller.rb]: No need for array.

File: app/controllers/lists_controller.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    end
+  end
+
+  def destroy
+    @list = List.find(params[:id])
+    title = @list.title
+
+    if @list.destroy
+      flash[:notice] = "\"#{title}\" was deleted successfully."
+      redirect_to [@list]
```
Comment (eliotsykes) [inline app/controllers/lists_controller.rb]: No need for array around `@list`

File: app/controllers/lists_controller.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    else
+      flash[:error] = "There was an error deleting the list. Please try again."
+      render :show
+    end
+  end
+
+  private
+
+  def list_params
```
Comment (eliotsykes) [inline app/controllers/lists_controller.rb:64]: Nice work, good, private method to aid reuse.

File: app/helpers/lists_helper.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+module ListsHelper
```
Comment (eliotsykes) [inline app/helpers/lists_helper.rb]: Delete blank helpers to reduce clutter.

File: app/views/lists/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,4 @@
+<%= simple_form_for [@list] do |f| %>
```
Comment (eliotsykes) [inline app/views/lists/new.html.erb]: Surrounding array isn't needed for `@list`.

File: spec/controllers/lists_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,40 @@
+require 'rails_helper'
+
+
+describe ListsController do
```
Comment (eliotsykes) [inline spec/controllers/lists_controller_spec.rb:4]: Currently the tests in this spec aren't using the `ListsController` action methods. Fix up the spec to use the controller action methods.

File: spec/features/sign_in_spec.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        click_link 'Sign In'
+      end
+      fill_in 'Email', with: user.email
+      fill_in 'Password', with: user.password
+
+      within 'form' do
+        click_button 'Sign in'
+      end
+
+      expect(current_path).to eq root_path
```
Comment (eliotsykes) [inline spec/features/sign_in_spec.rb:20]: Add expectations that prove the user is successfully logged in.

File: test/controllers/lists_controller_test.rb [hunk 11]
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
```
Comment (eliotsykes) [inline test/controllers/lists_controller_test.rb]: Delete the `blocitoff/test` directory as we're using `blocitoff/spec` to house our tests.

Consider using the `-T` option with `rails new ...` to skip creating the default `test/` directory on future projects, when you know that you're going to test with RSpec. e.g. `rails new mynewproject -T`

For a reminder of `rails new` options like `-T`, run `rails new --help`

**Comment by dannyvassallo** [comment]
@eliotsykes I think I got it going a little better now.

File: spec/controllers/lists_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+  include Devise::TestHelpers
+
+  before do
+    @user = create(:user)
+    sign_in @user
+  end
+
+  describe '#create' do
+    it "creates 2 lists" do
```
Comment (eliotsykes) [inline spec/controllers/lists_controller_spec.rb]: Description for `it` block needs updating

File: spec/controllers/lists_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+      post :create, list: { title: 'test', user: @user }
```
Comment (eliotsykes) [inline spec/controllers/lists_controller_spec.rb]: `user: @user` shouldn't be needed as a request param in any of the specs here.

File: spec/controllers/lists_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+      @list = List.last
+      expect(List.count).to eq(1)
+      expect(@list.title).to eq('test')
+      expect(response).to redirect_to(list_path(@list.id))
```
Comment (eliotsykes) [inline spec/controllers/lists_controller_spec.rb]: Minor tweak possible, `list_path(@list.id)` can be simplified to `list_path(@list)` in all specs

File: spec/features/sign_in_spec.rb [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+      expect(user.name).to eq 'Douglas Adams'
```
Comment (eliotsykes) [inline spec/features/sign_in_spec.rb]: Line 21 isn't yet checking if the user is signed in.

File: app/controllers/lists_controller.rb [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+    if @list.save
+      flash[:notice] = "List was saved."
+      redirect_to @list
+    else
+      flash[:error] = "There was an error saving the list. Please try again."
+      render :new
+    end
+  end
+
+  def update
```
Comment (eliotsykes) [inline app/controllers/lists_controller.rb:37]: Write a controller spec that has 2 users: `attacker` and `victim`. Log the `attacker` in and send an update request to update the `victim`'s list title.

Also write a controller spec where the logged-in `attacker` deletes the `victim`'s list.

Once these are failing as expected, then fix the controller by using scoped queries like:

``` ruby
@list = current_user.lists.find(params[:id])
```

**Comment by dannyvassallo** [comment]
@eliotsykes -- good to merge?

**Comment by eliotsykes** [comment]
Yep good to merge!

**Pull request merged**